perfmetrics/scripts/bigquery/experiments_gcsfuse_bq.py

```diff
@@ -0,0 +1,322 @@
+                                                                 Default value is bigquery.Client(project=project_id).
+  """
+
+  def __init__(self, project_id, dataset_id, bq_client=None):
+    if bq_client is None:
+      self.client = bigquery.Client(project=project_id)
+    else:
+      self.client = bq_client
+    self.project_id = project_id
+    self.dataset_id = dataset_id
+
+  @property
+  def dataset_ref(self):
+    """Gets the reference of the dataset
+
+    Returns:
+      google.cloud.bigquery.dataset.Dataset: The retrieved dataset object
+    """
+    return self.client.get_dataset(self.dataset_id)
+
+  def _get_table_from_table_id(self, table_id):
+    """Gets the table from BigQuery from table ID
+
+    Args:
+      table_id (str): String representing the ID or name of the table
+    Returns:
+      google.cloud.bigquery.table.Table: The table in BigQuery
+    """
+    table_ref = self.dataset_ref.table(table_id)
+    table = self.client.get_table(table_ref)
+    return table
+
+  def _execute_query(self, query) -> QueryJob:
+    """Executes the query in BigQuery and raises an exception if query
+       execution could not be completed.
+
+    Args:
+      query (str): Query that will be executed in BigQuery.
+
+    Raises:
+      Exception: If query execution failed.
+    """
+    job = self.client.query(query)
+    if job.errors:
+      for error in job.errors:
+        raise Exception(f"Error message: {error['message']}")
+    return job
+
+  def _check_if_config_valid(self, exp_config_id) -> bool:
+    """Checks if exp_config_id exists in the experiment_configuration table.
+
+    Args:
+      exp_config_id (str): An id that uniquely identifies an experiment
+
+    Returns:
+      bool: Returns true if exp_config_id exists, false otherwise.
+    """
+    query_check_if_config_valid = """
+      SELECT *
+      FROM `{}.{}.{}`
+      WHERE configuration_id = '{}'
+    """.format(self.project_id, self.dataset_id, constants.CONFIGURATION_TABLE_ID, exp_config_id)
+
+    job = self._execute_query(query_check_if_config_valid)
+    row_count = job.result().total_rows
+    if row_count:
+      return True
+    return False
+
+  def _insert_rows(self, table, rows_to_insert, table_id = None, config_id = None, start_time_build = None):
+    """Insert rows in table. If insertion of some nth row fails, delete (n-1) rows
+    that were inserted before and raise an exception
+
+    Args:
+      table (str): Table in which rows are being inserted
+      rows_to_insert (str): Rows to insert in the table
+      table_id (str): ID of table to which results are being uploaded
+      config_id (str): config_id of the experiment for which results are being uploaded
+      start_time_build (timestamp): Start epoch time of the build
+
+    Raises:
+      Exception: If some row insertion failed.
+    """
+    result = self.client.insert_rows(table, rows_to_insert)
```

Comment:
To answer your questions:
1. Correct. Result is an empty list if  insert_rows function inserts all rows and non-empty containing error information, if insertion of some row fails.
2. Some cases where insert_rows fails:
i) Authentication/Authorization.
ii) If the schema of the table doesn't match the structure of the rows being inserted.
iii) If the data being inserted violates any constraints defined on the table, such as data type mismatches or nullability constraints.
iv) If the insert operation exceeds the rate limits, it may fail with a rate limit error. This can occur if you are trying to insert a large number of rows in a short period. (https://cloud.google.com/bigquery/quotas#streaming_inserts)
v) Table or Dataset not found.
vi) Network or Connectivity issues

Network or Connectivity issues are out of control. The problem (iv) will not occur for our data size. The problem (i) will not occur as long as we use a service account associated with the gcs-fuse-test project. The problem (v) will not occur as long as we run the setup script once before any tests are run. As long as no changes are made to the extracted data and their types, problem (ii) and (iii) should also not occur.

Reply:
Thanks for the detailed analysis.

Regarding /2: Ideally, we don't expect errors to come, however we want to make sure if they occur, they don't leave the table in inconsistent state. Can failure happen inbetween say if one of the value in nth row has different type than the one specified in schema of tables ? if yes, then we should add try and except around the call to insert_rows and also delete the rows in case of failure.

Reply:
Yes, if some error does come, then I've added the code to delete the successfully inserted data and then raise an expection. Link to the delete query: https://github.com/GoogleCloudPlatform/gcsfuse/pull/1193/files/23e88afb5fb53a01de51ff74ea9b75791837473f#diff-d0e9344ff918825f350dde8a4717175b0d7bb137c67ca1514c099679db87b142R124:~:text=if%20config_id%3A-,query_check_if_row_exists,-%3D%20%22%22%22

Reply:
I meant would there be any case where the line "result = self.client.insert_rows(table, rows_to_insert)" itself fails (throws some exception) after inserting some rows. 
It is different from this line getting successful and returning errors in 'result'.